### PR TITLE
#2014 - Add callback url to GET and POST requests

### PR DESCRIPTION
**Pull request by MackHalliday**
- Add `callback_url` value to response in related GET and POST notifications request. 
- If `callback_url` in body of POST request, update value in DB. 
- Decided not to reuse `htttps_url` schema because of impact on POST to ServiceCallback

# Description

issue #2014 

## How Has This Been Tested?

Tested in Postman and confirmed that `callback_url` can be added to the Notification Object

### POST Notification SMS
Screenshot 2024-10-08 at 11 47 06 AM
Screenshot 2024-10-08 at 11 47 20 AM

### GET Notification 
Screenshot 2024-10-08 at 11 48 04 AM

### POST Notification Email
Screenshot 2024-10-08 at 11 48 42 AM

### POST Notification Validation Errors
Screenshot 2024-10-08 at 11 49 09 AM
Screenshot 2024-10-08 at 11 49 59 AM

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an [appropriate title](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/team_agreement.md#naming-prs): #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/pull_request_labels.md#vanotify-labels) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the [style guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/style_guide.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works. [Testing guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/testing_guide.md)
- [x] I have ensured the latest main is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] The ticket was moved into the DEV test column when I began testing this change

### CODE REVIEW

File: app/schema_validation/definitions.py [hunk 1]
```diff
@@ -48,10 +48,10 @@
 )
 
 https_url = {
```
Comment (MackHalliday) [inline app/schema_validation/definitions.py:50]: AC asked to reuse the `https_url` variable for the `callback_url` validation. This `https_url` variable to validate http urls is ONLY used for the callback url feature.

File: tests/app/v2/notifications/test_get_notifications.py
Comment (kalbfled) [inline tests/app/v2/notifications/test_get_notifications.py:1]: @k-macmillan Yikes!  I missed this during #1436.  Tests should be using the `sample_notification` fixture and not using `create_notification`.  I'm not asking Mack to change that for this ticket.

File: tests/app/v2/notifications/test_post_notifications.py [hunk 2]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+    assert resp_json['id'] == str(notification.id)
+    assert resp_json['content']['body'] == template.content.replace('(( Name))', 'Jo')
+    assert 'callback_url' not in resp_json or resp_json['callback_url'] is None
+
+
+@pytest.mark.parametrize(
+    'callback_url',
+    [
+        ('invalid-url', 'is not a valid URI.'),
+        ('htp://wrongformat.com', 'does not match ^https.*'),
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:432]: Did you mean to test "http"?

File: tests/app/v2/notifications/test_post_notifications.py [hunk 3]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+    notification = notifications[0]
+    assert notification.callback_url is None
+
+    assert resp_json['id'] == str(notification.id)
+    assert resp_json['content']['body'] == template.content.replace('(( Name))', 'Jo')
+    assert 'callback_url' not in resp_json or resp_json['callback_url'] is None
+
+
+@pytest.mark.parametrize(
+    'callback_url',
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:429]: Using the extra parameter here deletes code below.
```suggestion
    'callback_url, expected_error',
```

File: tests/app/v2/notifications/test_post_notifications.py [hunk 4]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+            'https://example.com/search?query=this_is_a_search_term_to_reach_exactly_256_charactersaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa&filter=type=all&status=active&sort=ascending&page=1234567890&session_token=abc123&tracking_id=unique_user_tracking_value',
+            'is too long',
+        ),
+    ],
+)
+def test_notification_returns_400_if_invalid_callback_url(
+    client,
+    sample_api_key,
+    sample_template,
+    callback_url,
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:444]:
```suggestion
    callback_url,
    expected_error,
```

File: tests/app/v2/notifications/test_post_notifications.py [hunk 5]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+):
+    template = sample_template(content='Hello (( Name))\nYour thing is due soon')
+    url = callback_url[0]
+    expected_error = callback_url[1]
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:448]:
```suggestion
```

File: tests/app/v2/notifications/test_post_notifications.py [hunk 6]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+
+    data = {
+        'phone_number': '+16502532222',
+        'template_id': str(template.id),
+        'personalisation': {' Name': 'Jo'},
+        'callback_url': url,
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:454]:
```suggestion
        'callback_url': callback_url,
```

File: tests/app/v2/notifications/test_post_notifications.py [hunk 7]
```diff
@@ -392,6 +395,75 @@ def test_notification_returns_400_and_for_schema_problems(
...
+    }
+
+    response = post_send_notification(client, sample_api_key(service=template.service), SMS_TYPE, data)
+
+    assert response.status_code == 400
+    assert response.headers['Content-type'] == 'application/json'
+    error_resp = response.get_json()
+
+    assert error_resp['status_code'] == 400
+    assert {'error': 'ValidationError', 'message': f'callback_url {url} {expected_error}'} in error_resp['errors']
```
Comment (kalbfled) [inline tests/app/v2/notifications/test_post_notifications.py:464]:
```suggestion
    assert {'error': 'ValidationError', 'message': f'callback_url {callback_url} {expected_error}'} in error_resp['errors']
```

**Review by kalbfled — CHANGES_REQUESTED** [review]

File: tests/app/v2/notifications/test_post_notifications.py (hunk 2 above)
Comment (MackHalliday) [reply tests/app/v2/notifications/test_post_notifications.py:432]: I put `htp` but `http` makes more sense.

File: tests/app/v2/notifications/test_post_notifications.py (hunk 3 above)
Comment (MackHalliday) [reply tests/app/v2/notifications/test_post_notifications.py:429]: Ah good call.

File: tests/app/v2/notifications/test_post_notifications.py (hunk 2 above)
Comment (MackHalliday) [reply tests/app/v2/notifications/test_post_notifications.py]: Updated.

**Review by kalbfled — APPROVED** [review]

File: tests/app/v2/notifications/test_post_notifications.py [hunk 8]
```diff
@@ -392,6 +395,76 @@ def test_notification_returns_400_and_for_schema_problems(
...
+    ).all()
+
+    assert len(notifications) == 1
+
+    notification = notifications[0]
+    assert notification.callback_url is None
+
+    assert resp_json['id'] == str(notification.id)
+    assert resp_json['content']['body'] == template.content.replace('(( Name))', 'Jo')
+    assert 'callback_url' not in resp_json or resp_json['callback_url'] is None
```
Comment (cris-oddball) [inline tests/app/v2/notifications/test_post_notifications.py:425]: `callback_url `should always be in the response? 
shouldn't we just check for `resp_json['callback_url'] is None` ?

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: tests/app/v2/notifications/test_post_notifications.py (hunk 8 above)
Comment (MackHalliday) [reply tests/app/v2/notifications/test_post_notifications.py]: Ops - Yes that should be just `resp_json['callback_url'] is None`

Updated.

**Review by kalbfled — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Pull request merged**